src/core/server/elasticsearch/elasticsearch_service.ts

```diff
@@ -0,0 +1,140 @@
+import { ElasticsearchConfig } from './elasticsearch_config';
+
+interface CoreClusterClients {
+  config: ElasticsearchConfig;
+  adminClient: ClusterClient;
+  dataClient: ClusterClient;
+}
+
+/** @internal */
+export interface ElasticsearchServiceStartContract {
+  // Required for the BWC only.
+  readonly bwc: {
+    readonly config: ElasticsearchConfig;
+  };
+
+  readonly apiVersion: ElasticsearchConfig['apiVersion'];
+  readonly requestTimeout: ElasticsearchConfig['requestTimeout'];
+  readonly shardTimeout: ElasticsearchConfig['shardTimeout'];
+
+  readonly createClient: (
+    type: string,
+    config?: Partial<ElasticsearchClientConfig>
+  ) => ClusterClient;
+  readonly adminClient: ClusterClient;
+  readonly dataClient: ClusterClient;
+}
+
+/** @internal */
+export class ElasticsearchService implements CoreService<ElasticsearchServiceStartContract> {
+  private readonly log: Logger;
+  private subscription?: Subscription;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.log = coreContext.logger.get('elasticsearch-service');
+  }
+
+  public async start(): Promise<ElasticsearchServiceStartContract> {
+    this.log.debug('Starting elasticsearch service');
+
+    const clients$ = this.coreContext.configService
+      .atPath('elasticsearch', ElasticsearchConfig)
+      .pipe(
+        filter(() => {
+          if (this.subscription !== undefined) {
+            this.log.error('Clients cannot be changed after they are created');
+            return false;
+          }
+
+          return true;
+        }),
+        switchMap(
+          config =>
+            new Observable<CoreClusterClients>(subscriber => {
+              this.log.debug(`Creating elasticsearch clients`);
+
+              const coreClients = {
+                config,
+                adminClient: this.createClusterClient('admin', config),
+                dataClient: this.createClusterClient('data', config),
+              };
+
+              subscriber.next(coreClients);
+
+              return () => {
+                this.log.debug(`Closing elasticsearch clients`);
+
+                coreClients.adminClient.close();
+                coreClients.dataClient.close();
+              };
+            })
+        ),
+        publishReplay(1)
+      ) as ConnectableObservable<CoreClusterClients>;
+
+    this.subscription = clients$.connect();
+
+    const { config: defaultConfig, adminClient, dataClient } = await clients$
+      .pipe(first())
+      .toPromise();
+
+    return {
+      bwc: { config: defaultConfig },
+
+      apiVersion: defaultConfig.apiVersion,
+      requestTimeout: defaultConfig.requestTimeout,
+      shardTimeout: defaultConfig.shardTimeout,
+
+      adminClient,
+      dataClient,
+
+      createClient: (type: string, clientConfig: Partial<ElasticsearchClientConfig> = {}) => {
+        return this.createClusterClient(type, {
+          ...(defaultConfig as ElasticsearchClientConfig),
```

Comment:
__note:__ I've noticed that in most of the places where we call `createCluster` we pass default ES config and I don't really see any reason to not use this one to fill the gaps if consumer doesn't specify something. Thoughts?